### PR TITLE
[AutoParallel] Add spmd rules for nonzero op

### PR DESCRIPTION
**Pull request by waliwali777**
### PR Category

Auto Parallel

### PR Types

New features

### Description

Add spmd rules for nonzero op
Pcard-67164

### CODE REVIEW

File: test/legacy_test/test_nonzero_op.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        np.random.seed(2023)
+        self.op_type = "where_index"
+        self.python_api = call_nonzero
+
+        self.inputs = self.create_inputs()
+        self.outputs = self.return_outputs()
+
+    def test_check_output(self):
+        self.check_output(
+            check_pir=True, check_auto_parallel=self.check_auto_parallel
```
Comment (gongshaotian) [inline test/legacy_test/test_nonzero_op.py]: nonero op的NonzeroOpInferSymbolicShape接口中直接使用了新符号，OpTest中的符号推导相关检查需要手动关闭。
check_symbol_infer=false

Comment (waliwali777) [reply test/legacy_test/test_nonzero_op.py]: Fixed, Thank you!

File: test/auto_parallel/spmd_rules/test_nonzero_rule.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zyfncg) [inline test/auto_parallel/spmd_rules/test_nonzero_rule.py]: 2023 -> 2024

**Review by zyfncg — DISMISSED** [review]

File: test/auto_parallel/spmd_rules/test_nonzero_rule.py (hunk 2 above)
Comment (waliwali777) [reply test/auto_parallel/spmd_rules/test_nonzero_rule.py]: Done

**Review by zyfncg — DISMISSED** [review]

**Review by liym27 — APPROVED** [review]
LGTM

**Review by zyfncg — APPROVED** [review]

**Pull request merged**